compiler/rustc_infer/src/infer/equate.rs

```diff
@@ -164,14 +165,18 @@ impl<'tcx> TypeRelation<'tcx> for Equate<'_, '_, 'tcx> {
             return Ok(a);
         }
 
-        if a.skip_binder().has_escaping_bound_vars() || b.skip_binder().has_escaping_bound_vars() {
-            self.fields.higher_ranked_sub(a, b, self.a_is_expected)?;
-            self.fields.higher_ranked_sub(b, a, self.a_is_expected)?;
+        let a = self.tcx().anonymize_bound_vars(a);
+        let b = self.tcx().anonymize_bound_vars(b);
+
+        if a.bound_vars() == b.bound_vars() {
```

Comment:
This is incompatible with lazy normalization, right? `for<'a> fn(Alias<'a>) == fn(u8)` would fail even if `Alias<'a> := u8`.

I think the solution is to do the equivalent of mutual `higher_ranked_sub` but in an invariant context.

Reply:
good point, yeah. Didn't think of that before 😅 🤔 

i would like to avoid the exponential blowup here due to calling using `higher_ranked_sub` twice per binder, but that feels like the easiest solution. Will think a bit more about it

Reply:
I guess we want the following ...

```diff
diff --git a/compiler/rustc_infer/src/infer/equate.rs b/compiler/rustc_infer/src/infer/equate.rs
index 5d929394eb0..3d26c6fa59c 100644
--- a/compiler/rustc_infer/src/infer/equate.rs
+++ b/compiler/rustc_infer/src/infer/equate.rs
@@ -165,8 +165,8 @@ fn binders<T>(
         }

         if a.skip_binder().has_escaping_bound_vars() || b.skip_binder().has_escaping_bound_vars() {
-            self.fields.higher_ranked_sub(a, b, self.a_is_expected)?;
-            self.fields.higher_ranked_sub(b, a, self.a_is_expected)?;
+            self.fields.higher_ranked_equate(a, b, self.a_is_expected)?;
+            self.fields.higher_ranked_equate(b, a, self.a_is_expected)?;
         } else {
             // Fast path for the common case.
             self.relate(a.skip_binder(), b.skip_binder())?;
```

`higher_ranked_equate` is the same as `higher_ranked_sub` but calls `equate` instead of `sub` and we stick with `higher_ranked_sub` for https://github.com/spastorino/rust/blob/f74f700952f105536446e415b8df8061bddfb25e/compiler/rustc_infer/src/infer/sub.rs#L200

Reply:
I've just applied said change. Proceeding with a new crater run.

Reply:
@spastorino I don't think a new crater run is needed for such change. Lazy normalization for projections is not a thing in the old solver and, for opaque types, we already disallow defining uses with late bound args.  The only occasion where this changes stable behavior is in coherence (because it doesn't do full region resolution and instead relies on the leak check)  but then it strictly accepts more code for which crater is not useful.

Reply:
@aliemjay anyway, I've removed the thing from the queue because it was just starting to execute and replaced it with the right one.